openshift-kube-apiserver/filters/deprecatedapirequest/crd_check.go

```diff
@@ -0,0 +1,59 @@
+package deprecatedapirequest
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apiextensionsinformers "k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+const CrdEstablished ContextKey = "DeprecatedApiRequestCrdEstablished"
+
+// WithCrdCheck adds a filter to the hander chain that set the CrdEstablished context key to true if the
+// deprecatedapirequests.apiserver.openshift.io CRD has been installed.
+func WithCrdCheck(handler http.Handler, config *rest.Config) http.Handler {
```

Comment:
I maybe missing the details here, but I was thinking of the following:
```
type DeprecatedAPIWatcher interface {
	IsEstablished() bool
	IsCacheSynced() bool

	IsRequestDeprecated(req *http.Request) bool
	Notify(req *http.Request)
}

func NewDeprecatedAPIWatcher() DeprecatedAPIWatcher {
	// initialization here
}

func WithDeprecatedAPIHandler(handler http.Handler, watcher DeprecatedAPIWatcher) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		if !(watcher.IsCacheSynced() && watcher.IsEstablished()) {
			handler.ServeHTTP(w, req)
			return
		}

		if watcher.IsRequestDeprecated(req) {
			watcher.Notify(req)
		}

		handler.ServeHTTP(w, req)
	})
}
```
This way we can:
- avoid using associating any keys in the request context. Right now, we have two keys `CachesSynced` and `CrdEstablished`. If we need an association, we can push the interface instance in to the request context. 
- handler chain does not have any initialization code.